lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+  ErrorHandler eh;  ///< error handler, collects messages
+  parser::CharBlock currentPosition_{
+      nullptr};  ///< current location in parse tree
+  ScopeProxy currentScope_{0};  ///< current scope in the model
+  std::vector<std::string> constructNames_;
+};
+
+template<typename A, typename B>
+bool InInclusiveScope(const A &scopes, B tail, const B &head) {
+  assert(HasScope(head));
+  assert(HasScope(tail));
+  while (HasScope(tail) && (tail != head)) {
+    tail = scopes[tail];
+  }
+  return tail == head;
+}
+
+ParseTreeAnalyzer LabelAnalysis(const parser::Program &program) {
+  ParseTreeAnalyzer analysis;
+  Walk(program, analysis);
+  return analysis;
+}
+
+template<typename A, typename B>
+inline bool InBody(const A &position, const B &pair) {
+  assert(pair.first.begin() < pair.second.begin());
+  return (position.begin() >= pair.first.begin()) &&
+      (position.begin() < pair.second.end());
+}
+
+template<typename A, typename B>
+LabelStmtInfo GetLabel(const A &labels, const B &label) {
+  const auto iter{labels.find(label)};
+  if (iter == labels.cend()) {
+    return {0u, nullptr, TargetStmtType{}};
+  } else {
+    return iter->second;
+  }
+}
+
+/// \brief Check branches into a <i>label-do-stmt</i>
+/// Relates to 11.1.7.3, loop activation
+template<typename A, typename B, typename C, typename D>
+inline void CheckBranchesIntoDoBody(const A &branches, const B &labels,
+    const C &scopes, const D &loopBodies, ErrorHandler &eh) {
+  for (const auto branch : branches) {
+    const auto &label{std::get<parser::Label>(branch)};
+    auto branchTarget{GetLabel(labels, label)};
+    if (HasScope(std::get<ScopeProxy>(branchTarget))) {
+      const auto &fromPosition{std::get<parser::CharBlock>(branch)};
+      const auto &toPosition{std::get<parser::CharBlock>(branchTarget)};
+      for (const auto body : loopBodies) {
+        if (!InBody(fromPosition, body) && InBody(toPosition, body)) {
+          // this is an error: branch into labeled DO body
+          if (isStrictF18) {
+            Report(eh, fromPosition,
+                parser::MessageFormattedText{
+                    "branch into '%s' from another scope"_err_en_US,
+                    body.first.ToString().c_str()});
+          } else {
+            Report(eh, fromPosition,
+                parser::MessageFormattedText{
+                    "branch into '%s' from another scope"_en_US,
+                    body.first.ToString().c_str()});
+          }
+        }
+      }
+    }
+  }
+}
+
+/// \brief Check that DO loops properly nest
+template<typename A>
+inline void CheckDoNesting(const A &loopBodies, ErrorHandler &eh) {
+  for (auto i1{loopBodies.cbegin()}; i1 != loopBodies.cend(); ++i1) {
+    const auto &v1{*i1};
+    for (auto i2{i1 + 1}; i2 != loopBodies.cend(); ++i2) {
+      const auto &v2{*i2};
+      assert(v1.first.begin() != v2.first.begin());
+      if ((v2.first.begin() < v1.second.end()) &&
```

Comment:
Might be more clear as something like `Overlap(v1, v2) && !Contains(v1, v2) && !Contains(v2, v1)`.